src/V3HierBlock.cpp

```diff
@@ -0,0 +1,436 @@
+//
+// There are 3 kinds of Verilator run.
+// a) To create ${prefix}_hier.mk (--hierarchical)
+// b) To create protect-lib for each hierarchical block (--hierarchical-child)
+// c) To load wrappers and Verilate the top module (... what primary flags?)
+//
+// Then user can build Verilated module as usual.
+//
+// Here is more detailed internal process.
+// 1) Parser adds AstPragmaType::HIER_BLOCK of AstPragma to modules
+//    that are marked with /*verilator hier_block*/ metacomment in Verilator run a).
+// 2) AstModule with HIER_BLOCK pragma is marked modp->hier_block(true)
+//    in V3LinkResolve.cpp during run a).
+// 3) In V3LinkCells.cpp, the following things are done during run b) and c).
+//    3-1) Delete the upper modules of the hierarchical block because the top module in run b) is
+//         hierarchical block, not the top module of run c).
+//    3-2) If the top module of the run b) or c) instantiates other hierarchical blocks that is
+//         parameterized,
+//         module and task names are renamed to the original name to be compatible with the
+//         hier module to be called.
+//
+//         Parameterized modules have unique name by V3Param.cpp. The unique name contains '__' and
+//         Verilator encodes '__' when loading such symbols.
+// 4) V3LinkDot.cpp checks dotted access across hierarchical block boundary.
+// 5) In V3Dead.cpp, some parameters of parameterized modules are protected not to be deleted even
+//    if the parameter is not referred. This protection is necessary to match step 6) below.
+// 6) In V3Param.cpp, use protect-lib wrapper of parameterized module made in b) and c).
+//    If a hierarchical block is a parameterized module and instantiated in multiple locations,
+//    all parameters must exactly match.
+// 7) In V3HierBlock.cpp, relationship among hierarchical blocks are checked in run a).
+//    (which block uses other blocks..)
+// 8) In V3EmitMk.cpp, ${prefix}_hier.mk is created in run a).
+//
+// There are two hidden command options.
+//   --hierarchical-child is added to Verilator run b).
+//   --hierarchical-block module_name,mangled_name,name0,value0,name1,value1,...
+//       module_name  :The original modulename
+//       mangled_name :Mangled name of parameterized modules (named in V3Param.cpp).
+//                     Same as module_name for non-parameterized hierarchical block.
+//       name         :The name of the parameter
+//       value        :Overridden value of the parameter
+//
+//       Used for b) and c).
+//       This options is repeated for all instantiating hierarchical blocks.
+
+#include <memory>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "V3Ast.h"
+#include "V3Error.h"
+#include "V3File.h"
+#include "V3HierBlock.h"
+#include "V3String.h"
+#include "V3Stats.h"
+
+typedef std::vector<std::pair<string, string> > StrGParams;
+
+static StrGParams stringifyParams(const V3HierBlock::GParams& gparams, bool forGOption) {
```

Comment:
```suggestion
static StrGParams stringifyParams(const V3HierBlock::GParams& gparams, bool forGOption) {
```

I prefer to avoid general global names even as statics unless they need to be global. Can you either prefix the function with the file e.g. v3HierStringifyParams, or better make these private method of V3HierBlock class.

Reply:
Done in e03034e5dafb9b461fee8c781bf05156b86bfbc2.
Made stringifyParams() as a member function.
hierCommandFileName() and writeCommonInputs() are used in two classes, so renamed to have `V3Hier` prefix.